### DISCUSSION TITLE
Data returned by  useLoaderData is always inferred as any

### DISCUSSION THREAD

**Discussion by micama** [remix-run/remix#10197] [Q&A]
Hello all

With a fresh project created using `npx create-remix@latest` the types of data returned from `useLoaderData<typeof loader>()` is not being inferred at all.

Am I missing something obvious here or is this being experienced by others as well?

I also tried to turn off the single fetch feature with the same result.

My expectation was for the data within the React component to be fully typed since the `json` shows the correct types.

Dependencies
![dependencies](https://github.com/user-attachments/assets/96503bab-e16f-4819-8eef-ee92a7da5e7f)

vite config
![vite-config](https://github.com/user-attachments/assets/02c0cc3e-2d3e-405a-a989-c5ab2274626d)

types in code
![json_method](https://github.com/user-attachments/assets/8d4391af-f4e1-4001-8c65-4c85bbb1ef68)
![loader_type](https://github.com/user-attachments/assets/2b8e7471-a287-4852-b935-8b2a0bec86d0)
![loaderdata_type](https://github.com/user-attachments/assets/fb593b88-8c0c-42eb-92a6-a895e16b534e)

Thanks

**Comment by micama** [comment]
Update:

If the loader function is not typed 
```ts
// No use of LoaderFunction type
export const loader = () => {
//....
}
```

then the type inference works.

So what is the best practice here?

**Reply by ss-joy** [reply]
Fun fact ,
I did not even realise that not using the `LoaderFunction` type would let me infer the correct type in the useLoader hook.
Currently,
the best solution is to not type the function using `LoaderFunction` and just to type the function parameter with `LoaderFunctionArgs`.

This way you get type inference in both the `loader` function and the `useLoaderData` hook.

**Comment by milamer** [comment]
Hey there, 

It appears that `LoaderFunction` and strict type safety don't interact well together. I recommend either not specifying a type for the function or using `LoaderFunctionArgs` if you need to type the arguments (like `request`, `params`, etc.).

[See Docs for more](https://remix.run/docs/en/main/route/loader#type-safety)

**Comment by micama** [comment]
This has nothing to do with the arguments though. 
As shown above the loader does not even use any arguments.

If the function cannot/should not be typified, this should be explicitly mentioned in the docs.
Better yet, the type should not be exported at all if that's the case. Otherwise it just confuses people.

Also, what function type should be used: normal/classic or arrow function? The documentation jumps between these two without giving any explanation.
It shouldn't make a difference, but regarding type inference maybe it does and should also be explicitly mentioned.

So my question remains, what is the best practice?

Thanks

**Comment by sergiodxa** [comment]
The `LoaderFunction` type is not compatible with type inference, you have to use `LoaderFunctionArgs` to type only the arguments.

**Comment by ss-joy** [comment]
Currently,
the best solution is to not type the function using `LoaderFunction` and just to type the function parameter with `LoaderFunctionArgs`.

This way you get type inference in both the **`loader`** function and the `**useLoaderData**` hook.
